langchain/llms/gpt4all.py

```diff
@@ -88,93 +88,65 @@ class GPT4All(LLM):
-            "top_p": self.top_p,
-            "temp": self.temp,
+            "n_ctx",
+            "n_predict",
+            "top_k",
+            "top_p",
+            "temp",
+            "n_batch",
+            "repeat_penalty",
+            "repeat_last_n",
+            "context_erase",
         }
 
-    def _gptj_default_params(self) -> Dict[str, Any]:
-        """Get the identifying parameters."""
+    def _default_params(self) -> Dict[str, Any]:
         return {
+            "n_ctx": self.n_ctx,
             "n_predict": self.n_predict,
-            "n_threads": self.n_threads,
             "top_k": self.top_k,
             "top_p": self.top_p,
             "temp": self.temp,
+            "n_batch": self.n_batch,
+            "repeat_penalty": self.repeat_penalty,
+            "repeat_last_n": self.repeat_last_n,
+            "context_erase": self.context_erase,
         }
 
-    @staticmethod
-    def _llama_param_names() -> Set[str]:
-        """Get the identifying parameters."""
-        return {
-            "seed",
-            "n_ctx",
-            "n_parts",
-            "f16_kv",
-            "logits_all",
-            "vocab_only",
-            "use_mlock",
-            "embedding",
-        }
-
-    @staticmethod
-    def _gptj_param_names() -> Set[str]:
-        """Get the identifying parameters."""
-        return set()
-
-    @staticmethod
-    def _model_param_names(backend: str) -> Set[str]:
-        if backend == "llama":
-            return GPT4All._llama_param_names()
-        else:
-            return GPT4All._gptj_param_names()
-
-    def _default_params(self) -> Dict[str, Any]:
-        if self.backend == "llama":
-            return self._llama_default_params()
-        else:
-            return self._gptj_default_params()
-
     @root_validator()
     def validate_environment(cls, values: Dict) -> Dict:
         """Validate that the python package exists in the environment."""
         try:
-            backend = values["backend"]
-            if backend == "llama":
-                from pygpt4all import GPT4All as GPT4AllModel
-            elif backend == "gptj":
-                from pygpt4all import GPT4All_J as GPT4AllModel
-            else:
-                raise ValueError(f"Incorrect gpt4all backend {cls.backend}")
-
-            model_kwargs = {
-                k: v
-                for k, v in values.items()
-                if k in GPT4All._model_param_names(backend)
-            }
+            from gpt4all import GPT4All as GPT4AllModel
+
+            full_path = values["model"]
+            model_path, delimiter, model_name = full_path.rpartition("/")
+            model_path += delimiter
+
             values["client"] = GPT4AllModel(
-                model_path=values["model"],
-                **model_kwargs,
+                model_name=model_name,
+                model_path=model_path,
+                model_type=values["backend"],
+                allow_download=False,
             )
```

Comment:
Rather than setting llama as a default backend, better left it as undefined, i.e.,
https://github.com/hwchase17/langchain/blob/23083e6a5c555e7c274b91658251edfd2eed0a82/langchain/llms/gpt4all.py#L31 to
```python
    backend: str = None
```
, and set the backend from the loaded model: 
```suggestion
            )
            values["backend"] = values["client"].model.model_type            
```
as of now, GPT4All can retrieve model from the model name (no need to set the model type from the user anymore)